book/src/dev/rfcs/0008-release-planning.md

```diff
@@ -0,0 +1,190 @@
+projects.  We would not be asking users to run autotools or `./configure`, just
+a one-step `cargo install`.  We also know that it's possible to reliably build
+Zebra on each platform with minimal additional steps, because we do so in our
+CI.
+
+In contrast to these downsides, distributing `zebra` through Cargo has a number
+of upsides.  First, because we distribute our libraries using crates.io, we
+already have to manage tooling for publishing to crates.io, so there's no
+additional work required to publish `zebrad` this way.  Second, we get a
+cross-platform update mechanism with no additional work, since `cargo install`
+will upgrade to the latest published version.  Third, we don't rely on any
+third parties to mediate the relationship between us and our users, so users
+can get updates as soon as we publish them.  Fourth, unlike a system package
+manager, we can pin exact hashes of every transitive dependency (via the
+`Cargo.lock`, which `cargo install` can be configured to respect).  Fifth,
+we're positioned to pick up (or contribute to) ecosystem-wide integrity
+improvements like a transparency log for `crates.io` or work on reproducible
+builds for Rust.
+
+This proposal is summarized above in the [guide-level
+explanation](#guide-level-explanation).
+
+## Release Processes
+
+The next question is what kind of release processes and automation we should
+use.  Here are two important priorities for these processes:
+
+1.  Reducing the friction of doing any individual release, allowing us to move
+    closer to a continuous deployment model;
+2.  Reducing the risk of error in the release process.
+
+These are roughly in order of priority but they're clearly related, since the
+more friction we have in the release process, the greater the risk of error,
+and the greater the risk of error, the more friction we require to prevent it.
+
+Automation helps to reduce friction and to reduce the risk of error, but
+although it helps enormously, it has declining returns after some point (in the
+sense that automating the final 5% of the work is significantly more complex
+and error-prone than automating the first 5% of the work).  So the challenge is
+to find the right "entry point" for the automated part of the system that
+strikes the right balance.
+
+One possibility is the following.  CD automation is triggered by pushing a new
+tag to the git repository.  Tags are specified as `crate-semver`, e.g.
+`zebra-network-3.2.8`, `zebrad-3.9.1`, etc.  When a new tag is pushed, the CD
+automation parses the tag to determine the crate name.  If it is `zebrad`, it
+builds a new Docker image and publishes the image and the crate.  Otherwise, it
+just publishes the crate.
+
+To publish a new version of any component crate, the process is:
+
+1.  Edit the `Cargo.toml` to increment the version number;
+2.  Update `crate/CHANGELOG.md` with a few human-readable sentences describing
+    changes since the last release (examples: [cl1], [cl2], [cl3]);
```

Comment:
I'd like to mostly handle updates to the CHANGELOG in PRs, as in, if I do a PR that adds multi non finalized chain handling that PR would be the one that adds a line about the chain handling to the unreleased section of the Changelog, that way our release process is just to audit the changelog and replace "Unreleased" with the correct version.

Reply:
If we have multiple PRs editing the changelog, then we'll get some merge conflicts.

Our strategies for resolving them could be:
* limit the number of changelog entries - don't make an entry for every PR
* make smaller feature PRs
* release more often
* add extra newlines between entries (seems trivial, but it works!)

Reply:
@yaahc Does handling updates to the CHANGELOG in PRs mean that there's a correspondence between PRs and changelog entries?  I'm not sure that correspondence exists, at least for human-readable changelogs.  For instance, a single feature could correspond to a single changelog entry but be composed of multiple PRs. The reason I suggested writing changelog entries as part of the release process is that that process seemed like the natural place to try to summarize the changes made over all PRs since the last release.

Having an "Unreleased" section at the top of the changelog seems like a good way to queue things, make sure they're not forgotten, and reduce the work required to do a release.

But I think it would still be good to keep this step and expand what it means to "audit the changelog" (which could include merging or editing changelog entries, etc.), because I think it's a good place to have human-in-the-loop checking.

Reply:
@teor2345 all of those strategies seem good, but particularly the last two (newlines between entries and frequent releases). :+1:

Reply:
> Does handling updates to the CHANGELOG in PRs mean that there's a correspondence between PRs and changelog entries?

not necessarily. Its reasonable for a PR to make no changes to the changelog, so in the multiple PR case you could have any one of those PRs update the changelog to summarize the change that they together made. Also I think it would be okay to at any time go and rewrite the content of the changelog, so if you feel like it's messy or there's some minor changes that shouldn't be there we could remove them, the point is just to have a live document that is well used.

Reply:
> But I think it would still be good to keep this step and expand what it means to "audit the changelog" (which could include merging or editing changelog entries, etc.), because I think it's a good place to have human-in-the-loop checking.

sounds great